### PR TITLE
Feature/debounce button

### PR DESCRIPTION
**Pull request by JeftaJochem**
## Description
In order to protect the hardware it is essential to debounce all inputs to the mbed. For now, only the emergency button is debounced. But the Class that was created can be used for any future mechanical input (button) that will be connected to the mbed. The on/off could be debounced later, this could be an issue for later. 

The emergency button was the top priority for this PR. A bouncing emergency button could rapidly switch the MOSFETs on and off and destroy them.  Also it is good practice to use debouncing on every mechanical input on any device. The debounce time [us] can be changed in the constructor, but for now the debounce time is set at one second(1000000 us). 

The software was successfully tested on an mbed.

## Changes
- Added button.cpp and Button.h
- Added debounceRead function in while loop of main.cpp of pdb
- Renamed `button` to `onOffButton`
- Changed `wait_ms()` to `wait_us()`

### CODE REVIEW

**Review by Olavhaasie — COMMENTED** [review]
Congrats on your first contribution :tada: Everythings seems ok, except for the code style. Fortunately, you can do this automatically. If you are using VS code, you can enable `Format on Save` in the settings. Finally, you have to fix the `src/pdb/main.cpp` and `src/standalone_pdb/main.cpp` files by using the `Control + Shift + I` key shortcut and then committing the changes.

Let me know if you need any more help :hand:

**Review by bjornminderman — CHANGES_REQUESTED** [review]
You've done it, your first PR! Overall it is looking very good. You might, however, also want to consider to alter the `standalone_pdb` environment in the same way you altered the `pdb` environment, since we might not switch to the PDB with EtherCAT in the very near future. Further (besides some small comments) nice job!

File: lib/button/Button.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+  this->debounceTimer.reset();
+  this->debounceTime = debounceTime;  // debouncetime in us
+}
+
+bool Button::debounceRead(bool initialButton)
+{  // Function to return debounced value
+  bool returnValue = initialButton;
+  bool readValue = this->input.read();
+  if (initialButton != readValue && this->debounceTimer.read_us() == 0)  // Is the button in the same state? has it
```
[the remark this answers is not in the document]
Comment (JorisWeeda) [reply lib/button/Button.cpp]: Could be! The code is highly intuitive so maybe just delete the comments, your call! :1st_place_medal:

**Review by Joris-JW — COMMENTED** [review]
Nice work!! :100:  As earlier suggested the coding style is a bit out of place but easy to fix with Olav his tips. As Bjorn requested the comments have to be checked again. In my opinion the code is nice and readable so the comments can maybe just be delete or, as we do in the hardware interface, do a general comment per function. Anyway, that's your call!

File: lib/StateMachine/StateMachine.cpp [hunk 2]
```diff
@@ -14,7 +14,7 @@ StateMachine::StateMachine()
 }
 
 // Updates the current state based on the button inputs
-void StateMachine::updateState(bool buttonState, bool masterOkState, bool shutdownAllowedState)
+void StateMachine::updateState(bool onOffButtonState, bool masterOkState, bool shutdownAllowedState)
 {
   int buttonTimeMs = this->onOffButtonTimer.read_ms();
   int ledTimeMs = this->ledTimer.read_ms();
```
Comment (JorisWeeda) [inline lib/StateMachine/StateMachine.cpp:20]: I would change these to ```read_us()``` to keep consistency.

File: lib/button/Button.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (JeftaJochem) [reply lib/button/Button.cpp]: I have added new comments, however I code is intuitive enough without comments.

**Comment by Olavhaasie** [comment]
@bjornminderman reminder :bell: some changes were made since your last review, so could you please take a look

File: src/pdb/main.cpp [hunk 3]
```diff
@@ -82,16 +83,18 @@ int main()
...
+  bool emergencyButtonState = emergencyButton.read();
+
   while (1)
   {
     // Update EtherCAT variables
     ecat.update();
 
     // Get inputs from digitalIns and I2C bus
-    bool buttonstate = button.read();
+    bool onOffButtonstate = onOffbutton.read();
```
Comment (bjornminderman) [inline src/pdb/main.cpp:94]: I would change `onOffButtonstate` to `onOffButtonState` for consistency reasons. Also the build fails on `onOffbutton.read()` which should be changed to `onOffButton.read()` 

```suggestion
    bool onOffButtonState = onOffButton.read();
```

File: src/pdb/main.cpp [hunk 4]
```diff
@@ -100,30 +103,32 @@ int main()
...
     uint8_t hvResetStates = hvControl.readAllReset();
 
+    emergencyButtonState = emergencyButton.debounceRead(emergencyButtonState);
+
     // Keep track of whether an EtherCAT master is present
     masterOnline = masterOnlineChecker.isOnline(mosi.masterOk);
 
     // Update system state
-    stateMachine.updateState(buttonstate, masterOnline, (bool)mosi.masterShutdownAllowed);
+    stateMachine.updateState(onOffButtonstate, masterOnline, (bool)mosi.masterShutdownAllowed);
```
Comment (bjornminderman) [inline src/pdb/main.cpp:112]:
```suggestion
    stateMachine.updateState(onOffButtonState, masterOnline, (bool)mosi.masterShutdownAllowed);
```

File: src/standalone_pdb/main.cpp [hunk 5]
```diff
@@ -8,11 +8,11 @@
...
 
 // On/off-related inputs/outputs
-DigitalOut buttonLed(LPC_BUTTON_LED, false);     // True means LED on
-DigitalOut mbedLed1(LPC_LED1, false);            // Shows same as button led
-DigitalIn button(LPC_BUTTON_PRESSED, PullDown);  // True means button pressed
-DigitalOut keepPDBOn(LPC_KEEP_PDB_ON, false);    // True means keep PDB on
-DigitalOut mbedLed4(LPC_LED4, false);            // Shows if in Shutdown state
+DigitalOut buttonLed(LPC_ONOFFBUTTON_LED, false);     // True means LED on
+DigitalOut mbedLed1(LPC_LED1, false);                 // Shows same as button led
+DigitalIn button(LPC_ONOFFBUTTON_PRESSED, PullDown);  // True means button pressed
```
Comment (bjornminderman) [inline src/standalone_pdb/main.cpp:13]: Maybe you can change this to onOffButton, just like you did for the PDB with EtherCAT. Don't forget to change this throuhout the entire file.

```suggestion
DigitalIn onOffButton(LPC_ONOFFBUTTON_PRESSED, PullDown);  // True means button pressed
```

**Review by bjornminderman — CHANGES_REQUESTED** [review]
Almost there, just a few little naming issues to tackle

File: src/standalone_pdb/main.cpp [hunk 6]
```diff
@@ -68,7 +68,7 @@ int main()
     uint8_t hvResetStates = hvControl.readAllReset();
 
     // Update system state
-    stateMachine.updateState(buttonstate, true, false);
+    stateMachine.updateState(OnOffbuttonState, true, false);
```
Comment (bjornminderman) [inline src/standalone_pdb/main.cpp:71]:
```suggestion
    stateMachine.updateState(OnOffButtonState, true, false);
```

**Review by bjornminderman — CHANGES_REQUESTED** [review]

**Review by Olavhaasie — APPROVED** [review]
@bjornminderman If you could take a final look :innocent:

**Review by bjornminderman — APPROVED** [review]

**Pull request merged**